Sidebar active tabs

PressJump opened the pull request:
Fixes #146 Implementing Sidebar Background Active Tabs

PressJump committed:
🚧 Individual Tabs

PressJump committed:
💬 Sidebar Active Tabs Settings

PressJump committed:
✨ Bind Keep Tabs Active to Preference

PressJump committed:
🔇 Remove Test Logs

PressJump committed:
🐛  Don't load again if active

PressJump committed:
🔧 Add KeepTabsActive to Preference

PressJump committed:
🩹  Persistence Fixes

PressJump committed:
Restart Tooltip & Checkbox Fix

PressJump committed:
Update main.js

PressJump committed:
🔥 Wrong Files

PressJump committed:
Merge 'alpha' into sidebar-active-tabs

PressJump commented:
## QA Requirements

- [ ] Sidebar

  - [ ] The sidebar exists
  - [ ] The settings and addons button are visible down the bottom
  - [ ] The bookmarks, history and synced tabs icons are in the sidebar
  - [ ] The sidebar can be resized properly
  - [ ] An [installed extension](https://addons.mozilla.org/en-US/firefox/addon/sidebar-note/?utm_source=addons.mozilla.org&utm_medium=referral&utm_content=search) will appear in the sidebar
  - [ ] Sidebar-Tabs is Loaded and has a button visible down the bottom
  - [ ] Sidebar-Tabs can add a sidebar item to the sidebar
  - [ ] Sidebar-Tabs can remove a sidebar item to the sidebar



- [ ] Extensions

  - [ ] Tabliss is loading correctly
  - [ ] Tabliss does not have the webextention symbol in the title bar
  - [ ] uBlock origin is loading correctly
  - [ ] QRCode generator is loading correctly

- [ ] About dialog

  - [ ] The new about dialog is present
  - [ ] The dialog has not significantly changed

- [ ] Pocket

  - [ ] Pocket is not present in the browser

- [ ] Preferences

  - [ ] All tabs in preferences open
  - [ ] The "More from {Vendor}" option is not present
  - [ ] The sidebar preferences are present and work as expected

- [ ] Welcome experience

  - [ ] The import button opens the import wizard
  - [ ] The theme picker is present and works as expected
  - [ ] The search engine picker is present and works as expected

- [ ] Pulse theme

  - [ ] The tabs appear to be consistent

- [ ] Mozilla theme

  - [ ] The sidebar tabs are correct

- [ ] Useragent
  - [ ] The user agent is reporting as Mozilla Firefox

PressJump commented:
Currently browsers do not keep up with new sidebar sizing on 107, and removed shortcuts still exist in the background. With changes coming with Sidebar Active Tabs, the QA checks will need an update too. In draft for now due to this.

PressJump committed:
💬 QA Update

PressJump committed:
💬 ⚰️ Remove Unneeded QA Check

PressJump committed:
Fix Promise

PressJump committed:
Fix Load for Preference Disabled

PressJump commented:
Removed shortcuts now get removed after restart of the browser. This is the case with every other loaded tab as well. Persistence is only kept when keepTabsActive is set to false.

PressJump commented:
Welcome options for sidebar will be available on next PR.

PressJump committed:
✨ Add Sidebar Options to Welcome

PressJump committed:
Revert "✨ Add Sidebar Options to Welcome"

This reverts commit 3e8797c53e8196ad611b981ca66770338b484fb2.

trickypr commented on .github/PULL_REQUEST_TEMPLATE/ff_update.md:
```diff
@@ -14,7 +14,8 @@ This bumps Firefox up to `{NEW_VERSION}` for security and performance reasons. R
   - [ ] Sidebar-Tabs can add a sidebar item to the sidebar
   - [ ] Sidebar-Tabs can remove a sidebar item to the sidebar
-
+  - [ ] The sidebar keeps tabs active while closed when the preference `pulse.sidebar.keeptabsactive.enabled` is set to true
```
We do not care about preference keys, we only care about user facing options.

trickypr commented on src/browser/base/content/browser-sidebar-js.patch:
```diff
@@ -134,7 +144,37 @@ index f09e37fc42f7754f7f3687884ead25124c1b2076..e1833bbfd8a8ed7da767349f886d554f
++    
++    //check if element by id sidebar exists to bool var
++
++    var sidebar = document.getElementById("sidebar");
```
Is there a reason why you aren't using `this.browser` here?

trickypr commented on src/browser/base/content/browser-sidebar-js.patch:
```diff
@@ -84,16 +84,26 @@ index f09e37fc42f7754f7f3687884ead25124c1b2076..e1833bbfd8a8ed7da767349f886d554f
++    if (this._keepTabsActive != null) {
++      return this._keepTabsActive;
++    }
++    return this._keepTabsActive = Services.prefs.getBoolPref("pulse.sidebar.keeptabsactive.enabled", true);
```
Why is there not an observer to detect when this preference changes and reset everything? Have I missed it?

trickypr commented on src/browser/locales/en-US/browser/preferences/preferences-ftl.patch:
```diff
@@ -11,6 +11,8 @@ index b10403800c7aafe97471474f6bddb6098ce21a82..b39593816149a71a1e6b40626e304e36
 +    .label = Show available extensions on the sidebar
++preferences-sidebar-active-tabs = 
++    .label = Keep sidebar tabs active even when closed
++preferences-sidebar-active-tabs-tooltip = When enabled, sidebar tabs will remain active even when closed. When disabled, sidebar tabs will be closed when you click on another tab. <a data-l10n-name="restart-browser">Restart { -brand-short-name }</a>
```
Preference changes do not require restarts. If they do, you have written incorrect code.

trickypr commented on src/browser/components/preferences/main-inc-xhtml.patch:
```diff
@@ -13,6 +13,13 @@ index 3d4035690d38f4d851cc86749ebeec7b7ca9aaf2..e9136a52427d7f3ada87803373f16c20
++  <checkbox id="sidebarKeepTabsActive"
++    preference="pulse.sidebar.keeptabsactive.enabled"
++    data-l10n-id="preferences-sidebar-active-tabs"/>
++  <description id="sidebarKeepTabsActiveTooltip"
```
This is a description (mozilla calls it a description or footer depending on the circumstance), not a tooltip.

trickypr reviewed (changes requested):
This will not be merged if it requires a browser restart to work correctly. Either:
a) Remove the preference
b) Add a preference observer and fix whatever ends up being broken that requires a browser restart

trickypr commented on src/browser/base/content/browser-sidebar-js.patch:
```diff
@@ -165,7 +205,89 @@ index f09e37fc42f7754f7f3687884ead25124c1b2076..e1833bbfd8a8ed7da767349f886d554f
++      console.log("returned");
++      return;
++    }
++    console.log("it went on");
```
Also, you still have debug console.log statements here

PressJump replied:
Also left overs from testing which I forgot to wipe.

trickypr commented on src/browser/base/content/browser-sidebar-js.patch:
```diff
@@ -134,7 +144,37 @@ index f09e37fc42f7754f7f3687884ead25124c1b2076..e1833bbfd8a8ed7da767349f886d554f
+ 
+     this._initDeferred.resolve();
++    
++    //check if element by id sidebar exists to bool var
```
What is this comment? It doesn't appear to add much of value

PressJump replied:
Probably remnant of testing and etc that I forgot to wipe, not ready to review yet.

PressJump commented:
Changed to draft until changes made, I am not too sure why it was not on draft already after last review.

PressJump committed:
Observer

PressJump committed:
🔇 this.Browser & Remove Logs

PressJump committed:
💡 Update comment with relevant info

PressJump committed:
🌐 Change Lang Tags and Remove Restart Button

PressJump commented:
Bug where a sidebar item gets launched into the built in icons after disabling the keep sidebar items active. 
![image](https://user-images.githubusercontent.com/40802367/209425500-1ca356a3-e431-48c0-bce7-229f15d6f592.png)

PressJump committed:
Sidebar Seperator Fix

PressJump committed:
🔀 Merge branch 'alpha' into sidebar-active-tabs

PressJump commented:
Preference now relies on observer, reloading the sidebar and removing any left over content and changes have been updated to sync with alpha.

PressJump committed:
Change from preference key to settings item name

trickypr commented on src/browser/components/preferences/main-js.patch:
```diff
@@ -1,15 +1,17 @@
 +
 +  // Pulse browser: custom settings
 +  { id: 'pulse.sidebar.enabled', type: 'bool' },
++  { id: 'pulse.sidebar.extensions.enabled', type: 'bool' },
```
Why is the extension preference back?

trickypr committed:
⚡️ Fix memory leak

Also, you forgot to run `./mach lint --fix [file]`, so the formatting
has changed to be correct.

PressJump committed:
🔥  Weird, must have been a merge issue

PressJump reviewed (commented):
Yea sorry, forgot to lint.
